wa/instruments/perf.py

```diff
@@ -13,126 +13,220 @@
+                    'options': options,
+                    'args': {
+                        'command': 'sleep',
+                        'args': 1000,
+                    },
+                    'stderr': '&1',
+                    'stdout': 'stat.out',
+                }
+                for label, options in zip(self.labels, self.optionstring)
+            })
+        # TODO: else, detect if events or optionstring were passed and error?
+
         self.collector = PerfCollector(self.target,
-                                       self.events,
-                                       self.optionstring,
-                                       self.labels,
-                                       self.force_install)
+                                       self.force_install,
+                                       self.pre_commands,
+                                       self.commands,
+                                       self.post_commands)
 
     def setup(self, context):
+        # pylint: disable=unused-argument
         self.collector.reset()
 
     def start(self, context):
+        # pylint: disable=unused-argument
         self.collector.start()
 
     def stop(self, context):
+        # pylint: disable=unused-argument
         self.collector.stop()
 
     def update_output(self, context):
-        self.logger.info('Extracting reports from target...')
         outdir = os.path.join(context.output_directory, 'perf')
-        self.collector.get_trace(outdir)
-
-        for host_file in os.listdir(outdir):
-            label = host_file.split('.out')[0]
-            host_file_path = os.path.join(outdir, host_file)
-            context.add_artifact(label, host_file_path, 'raw')
-            with open(host_file_path) as fh:
-                in_results_section = False
-                for line in fh:
-                    if 'Performance counter stats' in line:
-                        in_results_section = True
-                        next(fh)  # skip the following blank line
-                    if in_results_section:
-                        if not line.strip():  # blank line
-                            in_results_section = False
-                            break
-                        else:
-                            line = line.split('#')[0]  # comment
-                            match = PERF_COUNT_REGEX.search(line)
-                            if match:
-                                classifiers = {}
-                                cpu = match.group(1)
-                                if cpu is not None:
-                                    classifiers['cpu'] = int(cpu.replace('CPU', ''))
-                                count = int(match.group(2))
-                                metric = '{}_{}'.format(label, match.group(3))
-                                context.add_metric(metric, count, classifiers=classifiers)
+        self.collector.get_traces(outdir)
+        # HUGE TODO: add parsers for supported post_commands
```

Comment:
Now that we support any `perf` command, what are the expectations for the output parser? I understand that WA captures those metrics into a JSON output file. Although these are adapted for cases having limited numbers of metrics (such as `perf stat`, as was done previously), it might not be ideal for large tables similar to what `perf report` may output. Are we expecting this instrument to parse this table into those JSON results? it's feasible and seems to meet what was done previously but is probably not the most practical approach.

Reply:
WA supports two kinds of outputs from  plugins: metrics and artifiacts. Metrics are individual values intended to be viewed and processed individually, artefacts are external entities (usually files) that are intended to be be handed as a monolithic unit. 

When you have something that generates a large number of values you have two options: add those values as individual metrics, or write them to a file (e.g. as a csv table) and add that file as an artefact. 

Which of these approaches you chose should be a function of how those values are intended to be used, rather than their number. If you expect to be able to query and compare individual values (e.g. across runs, or across configurations within a single run), they should be added as metrics. If the values only make sense when viewed as a group (e.g. it's a trace, a time series, etc), then they should written out into an artefact.